File: 2. Solved Problems/20. Valid Parentheses/20-valid-parentheses.java
//* https://leetcode.com/problems/valid-parentheses/description/

class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();

        for(int i = 0; i < s.length(); ++i){
            char ch = s.charAt(i);

            if(ch == '(' || ch == '{' || ch == '['){
                st.push(ch);
            }

            else{
                if(st.isEmpty()){
                    return false;
                }

                else{
                    char c = st.peek();
                    if(ch == ')' && c == '(' || ch == '}' && c == '{' || ch == ']' && c == '['){
                        st.pop();
                    }
                    else{
                        return false;
                    }
                }
            }
        }

        if(!st.isEmpty()){
            return false;
        }
        else{
            return true;
        }
    }
}

File: 2. Solved Problems/20. Valid Parentheses/20-valid-parentheses.cpp
//* https://leetcode.com/problems/valid-parentheses/description/
//* https://leetcode.com/problems/valid-parentheses/submissions/896420852/

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        // map<char,char> mp;
        // mp[')'] = '(';
        // mp[']'] = '[';
        // mp['}'] = '{';
        for(auto& ch : s){
            if(ch == '(' || ch == '[' || ch == '{'){
                st.push(ch);
            }
            else if(st.size() > 0){
                char c = st.top();

                // if(mp[ch] == st.top())
                if((c=='(' && ch == ')') || (c=='[' && ch == ']') || (c=='{' && ch == '}'))
                {
                    st.pop();
                } else{
                    return false;
                }
            }
            else{
                return false;
            }
        }

        if(st.size() > 0){
            return false;
        } else{
            return true;
        }
    }
};